### DISCUSSION TITLE
[nodemailer] `Property 'pass' does not exist on type 'AuthenticationTypeOAuth2'` after type guarding `type` property

### DISCUSSION THREAD

**Discussion by b-novikov-ipersonality** [DefinitelyTyped/DefinitelyTyped#60164] [Issues with a @types package]
Hi guys. Check out [codesandbox](https://codesandbox.io/s/priceless-meitner-nooi3b?file=/src/MailService.ts:1122-1244) to see what I'm talking about.  

Long story short, after type guarding that `type` property on `auth` object is of type `AuthenticationTypeLogin`, which means it has either value `"login"` or `undefined`, with both values supporting credentials with login and password, typescript yells that `pass` property does not exist on `AuthenticationTypeOAuth2` type. 

As I understand it happens because both `AuthenticationTypeOAuth2` and `AuthenticationTypeLogin` can have `type` value of `undefined` and typescript doesn't understand which type I mean. 

How can I solve it? Is explicitly passing auth type to constructor is the only option to avoid error?

[a comment by an automated account is not shown]

**Comment by Kurumi30** [comment]
> Hi guys. Check out [codesandbox](https://codesandbox.io/s/priceless-meitner-nooi3b?file=/src/MailService.ts:1122-1244) to see what I'm talking about.
> 
> Long story short, after type guarding that `type` property on `auth` object is of type `AuthenticationTypeLogin`, which means it has either value `"login"` or `undefined`, with both values supporting credentials with login and password, typescript yells that `pass` property does not exist on `AuthenticationTypeOAuth2` type.
> 
> As I understand it happens because both `AuthenticationTypeOAuth2` and `AuthenticationTypeLogin` can have `type` value of `undefined` and typescript doesn't understand which type I mean.
> 
> How can I solve it? Is explicitly passing auth type to constructor is the only option to avoid error?

I fixed the issue by using a type assertion to ensure that auth is treated as a Credentials object. Here’s the code I used:
```ts
import { Credentials } from "nodemailer/lib/smtp-connection"
// your code
if (!(auth as Credentials)?.user || !(auth as Credentials)?.pass) {
  throw new Error("user or password was not provided");
}
```
The key part is the type assertion (auth as Credentials), which helps TypeScript understand the shape of auth and avoid type errors.

Hope this helps, and I apologize for my poor English.
